### PR TITLE
dbworker: Mark errored records "failed" if retries are exhausted

### PR DESCRIPTION
**Pull request by mrnugget**
This fixes https://github.com/sourcegraph/sourcegraph/issues/16046 by implementing the idea in https://github.com/sourcegraph/sourcegraph/issues/16046#issuecomment-737173357.

It changes the `MarkErrored` method to either mark a record as errored or as failed, depending on whether it has exhausted all its retries or not.

Thoughts?

### CODE REVIEW

File: internal/workerutil/dbworker/store/store.go [hunk 1]
```diff
@@ -498,11 +501,22 @@ RETURNING {id}
...
 // with an error will not be updated. This method returns a boolean flag indicating if the record was updated.
 func (s *store) MarkFailed(ctx context.Context, id int, failureMessage string) (bool, error) {
-	q := s.formatQuery(markErroredOrFailedQuery, quote(s.options.TableName), "failed", failureMessage, id)
+	q := s.formatQuery(markFailedQuery, quote(s.options.TableName), failureMessage, id)
 	_, ok, err := basestore.ScanFirstInt(s.Query(ctx, q))
 	return ok, err
 }
 
+const markFailedQuery = `
+-- source: internal/workerutil/store.go:MarkErrored|MarkFailed
```
Comment (efritz) [inline internal/workerutil/dbworker/store/store.go:510]:
```suggestion
-- source: internal/workerutil/store.go:MarkFailed
```

File: internal/workerutil/dbworker/store/store.go [hunk 2]
```diff
@@ -481,15 +481,18 @@ RETURNING {id}
...
 func (s *store) MarkErrored(ctx context.Context, id int, failureMessage string) (bool, error) {
-	q := s.formatQuery(markErroredOrFailedQuery, quote(s.options.TableName), "errored", failureMessage, id)
+	q := s.formatQuery(markErroredQuery, quote(s.options.TableName), s.options.MaxNumRetries, failureMessage, id)
 	_, ok, err := basestore.ScanFirstInt(s.Query(ctx, q))
 	return ok, err
 }
 
-const markErroredOrFailedQuery = `
+const markErroredQuery = `
 -- source: internal/workerutil/store.go:MarkErrored|MarkFailed
```
Comment (efritz) [inline internal/workerutil/dbworker/store/store.go:490]:
```suggestion
-- source: internal/workerutil/store.go:MarkErrored
```

File: internal/workerutil/dbworker/store/store.go [hunk 3]
```diff
@@ -498,11 +501,22 @@ RETURNING {id}
...
+	q := s.formatQuery(markFailedQuery, quote(s.options.TableName), failureMessage, id)
 	_, ok, err := basestore.ScanFirstInt(s.Query(ctx, q))
 	return ok, err
 }
 
+const markFailedQuery = `
+-- source: internal/workerutil/store.go:MarkErrored|MarkFailed
+UPDATE %s
+SET {state} = 'failed',
+    {finished_at} = clock_timestamp(),
```
Comment (efritz) [inline internal/workerutil/dbworker/store/store.go:513]:
```suggestion
	{finished_at} = clock_timestamp(),
```

File: internal/workerutil/dbworker/store/store.go [hunk 4]
```diff
@@ -481,15 +481,18 @@ RETURNING {id}
...
 	return ok, err
 }
 
-const markErroredOrFailedQuery = `
+const markErroredQuery = `
 -- source: internal/workerutil/store.go:MarkErrored|MarkFailed
 UPDATE %s
-SET {state} = %s, {finished_at} = clock_timestamp(), {failure_message} = %s, {num_failures} = {num_failures} + 1
+SET {state} = CASE WHEN {num_failures} + 1 = %d THEN 'failed' ELSE 'errored' END,
+    {finished_at} = clock_timestamp(),
```
Comment (efritz) [inline internal/workerutil/dbworker/store/store.go:493]:
```suggestion
	{finished_at} = clock_timestamp(),
```

**Review by efritz — APPROVED** [review]

**Comment by efritz** [comment]
This _may_ cause a lot of the frontend/graphql resolvers to fail if we're not converting the new 'failed' state correctly. Could you either confirm that failed records are ok or wait for me to validate before merge?

**Comment by mrnugget** [comment]
> Could you either confirm that failed records are ok or wait for me to validate before merge?

I can do this, but I'm not sure what the best strategy is to, as they say in Pokemon, _catch them all_?

**Review by LawnGnome — APPROVED** [review]
LGTM from a campaigns perspective, but I obviously don't have @efritz's broader context here.

**Review by efritz — CHANGES_REQUESTED** [review]
This breaks auto-indexing (and LSIF processing, I think) due to the CASE statement in the markFailed query. Let me explore this a bit.

We might have to get rid of the enum types.

**Comment by efritz** [comment]
#16384 and #16386 should be enough to preserve behavior, although #16385 is also nasty in that it hid the error message that made the markFailed query silently lock auto-index records.

Once these three get merged in let me try to reproduce some issues one more time, then this :baby: :eyes: 🎉 !

**Comment by mrnugget** [comment]
Did you see my comment here? https://github.com/sourcegraph/sourcegraph/issues/16046#issuecomment-737248732 That might also be viable and, in contrast to this, would contain this behaviour to just campaigns code, no changes needed in other consumer code of `dbworker`.

Looks like you already did the work, but in maybe we should reconsider if this is another option?

**Comment by efritz** [comment]
> That might also be viable and, in contrast to this, would contain this behaviour to just campaigns code, no changes needed in other consumer code of dbworker.

I think the changes here are good and also general enough to apply to all consumers. I'll want to be able to use markFailed &co in the worker and the executor at some point (if gitserver is down of course we want to retry!)

**Review by efritz — APPROVED** [review]
Merged in those changes to fix usage in codeintel - LGTM!

🚢

**Review by chrispine — COMMENTED** [review]
LGTM

**Pull request merged**